File: app/src/gui/tabs/search_tab.py
import os
import platform
import subprocess

from typing import Dict, Any, List, Optional
from PySide6.QtGui import QPixmap
from PySide6.QtCore import Qt, QTimer
from PySide6.QtWidgets import (
    QFormLayout, QHBoxLayout, QVBoxLayout, QGridLayout,
    QScrollArea, QGroupBox, QApplication, 
    QLineEdit, QPushButton, QComboBox,
    QWidget, QLabel, QMessageBox
)
from .base_tab import BaseTab
from ..components import OptionalField
from ..components import ImagePreviewWindow
from ..styles.style import apply_shadow_effect
from ...utils.definitions import SUPPORTED_IMG_FORMATS


class SearchTab(BaseTab):
    def __init__(self, db_tab_ref, dropdown=True):
        super().__init__()
        # Reference to the main DatabaseTab to access the self.db connection object
        self.db_tab_ref = db_tab_ref
        self.dropdown = dropdown
        self.result_widgets = []
        self.open_preview_windows = [] # Track open preview windows
        self.selected_formats = set() # NEW
        
        layout = QVBoxLayout(self)
        
        # --- Search Criteria ---
        search_group = QGroupBox("Search Database")
        
        form_layout = QFormLayout(search_group)
        form_layout.setContentsMargins(10, 20, 10, 10)
        
        # Group name
        self.group_combo = QComboBox()
        self.group_combo.setEditable(True)
        self.group_combo.setPlaceholderText("e.g., Summer Trip (Optional)")
        form_layout.addRow("Group name:", self.group_combo)
        
        # Subgroup Name
        self.subgroup_combo = QComboBox()
        self.subgroup_combo.setEditable(True)
        self.subgroup_combo.setPlaceholderText("e.g., Beach Photos (Optional)")
        form_layout.addRow("Subgroup name:", self.subgroup_combo)
        
        # Filename
        self.filename_edit = QLineEdit()
        self.filename_edit.setPlaceholderText("e.g., *.png, img_001, etc (Optional)")
        self.filename_field = OptionalField("Filename pattern", self.filename_edit, start_open=False)
        form_layout.addRow(self.filename_field)
        
        # --- NEW: Input formats (copied from Convert.py) ---
        if self.dropdown:
            formats_layout = QVBoxLayout()
            btn_layout = QHBoxLayout()
            self.format_buttons = {}
            for fmt in SUPPORTED_IMG_FORMATS:
                btn = QPushButton(fmt)
                btn.setCheckable(True)
                btn.setStyleSheet("QPushButton:hover { background-color: #3498db; }")
                apply_shadow_effect(btn, color_hex="#000000", radius=8, x_offset=0, y_offset=3)
                btn.clicked.connect(lambda checked, f=fmt: self.toggle_format(f, checked))
                btn_layout.addWidget(btn)
                self.format_buttons[fmt] = btn
            formats_layout.addLayout(btn_layout)

            all_btn_layout = QHBoxLayout()
            self.btn_add_all = QPushButton("Add All")
            self.btn_add_all.setStyleSheet("background-color: green; color: white;")
            apply_shadow_effect(self.btn_add_all, color_hex="#000000", radius=8, x_offset=0, y_offset=3)
            self.btn_add_all.clicked.connect(self.add_all_formats)
            self.btn_remove_all = QPushButton("Remove All")
            self.btn_remove_all.setStyleSheet("background-color: red; color: white;")
            apply_shadow_effect(self.btn_remove_all, color_hex="#000000", radius=8, x_offset=0, y_offset=3)
            self.btn_remove_all.clicked.connect(self.remove_all_formats)
            all_btn_layout.addWidget(self.btn_add_all)
            all_btn_layout.addWidget(self.btn_remove_all)
            formats_layout.addLayout(all_btn_layout)

            formats_container = QWidget()
            formats_container.setLayout(formats_layout)
            self.formats_field = OptionalField("Input formats", formats_container, start_open=False)
            form_layout.addRow(self.formats_field)
        else:
            self.selected_formats = None
            self.input_formats_edit = QLineEdit()
            self.input_formats_edit.setPlaceholderText("e.g. jpg png gif (optional)")
            form_layout.addRow("Input formats:", self.input_formats_edit)
        # --- END NEW ---

        # Tags
        self.tags_edit = QLineEdit()
        self.tags_edit.setPlaceholderText("tag1, tag2, tag3... (comma-separated, optional)")
        form_layout.addRow("Tags:", self.tags_edit)
        
        layout.addWidget(search_group)
        
        # Search button
        self.search_button = QPushButton("Search Database")
        self.search_button.setStyleSheet("""
            QPushButton {
                background: qlineargradient(x1:0, y1:0, x2:0, y2:1,
                    stop:0 #667eea, stop:1 #764ba2);
                color: white; font-weight: bold; font-size: 16px;
                padding: 14px; border-radius: 10px; min-height: 44px;
            }
            QPushButton:hover { background: qlineargradient(x1:0, y1:0, x2:0, y2:1,
                stop:0 #764ba2, stop:1 #667eea); }
            QPushButton:disabled { background: #4f545c; color: #a0a0a0; }
            QPushButton:pressed { background: #5a67d8; }
        """)
        apply_shadow_effect(self.search_button, color_hex="#000000", radius=8, x_offset=0, y_offset=3)
        self.search_button.clicked.connect(self.perform_search)
        layout.addWidget(self.search_button)
        
        # Connect Enter key to search
        self.group_combo.lineEdit().returnPressed.connect(self.search_button.click)
        self.subgroup_combo.lineEdit().returnPressed.connect(self.search_button.click)
        self.filename_edit.returnPressed.connect(self.search_button.click)
        self.tags_edit.returnPressed.connect(self.search_button.click)
        if not self.dropdown:
            self.input_formats_edit.returnPressed.connect(self.search_button.click)
        
        # Results area
        results_label = QLabel("Search Results:")
        results_label.setStyleSheet("font-weight: bold; font-size: 12px; margin-top: 10px;")
        layout.addWidget(results_label)
        
        self.results_count_label = QLabel("Not connected to database.")
        self.results_count_label.setStyleSheet("color: #aaa; font-style: italic;")
        layout.addWidget(self.results_count_label)
        
        # Scrollable area for image results
        self.results_scroll = QScrollArea()
        self.results_scroll.setWidgetResizable(True)
        self.results_scroll.setMinimumHeight(300)
        
        self.results_widget = QWidget()
        self.results_layout = QGridLayout(self.results_widget)
        self.results_layout.setAlignment(Qt.AlignTop | Qt.AlignHCenter)
        self.results_scroll.setWidget(self.results_widget)
        
        layout.addWidget(self.results_scroll)
        
        self.setLayout(layout)

        # Update enabled state based on DB connection
        self.update_search_button_state()

    # --- NEW: Helper methods copied from Convert.py ---
    def toggle_format(self, fmt, checked):
        if checked:
            self.selected_formats.add(fmt)
            self.format_buttons[fmt].setStyleSheet("""
                QPushButton:checked { background-color: #3320b5; color: white; }
                QPushButton:hover { background-color: #00838a; }
            """)
            apply_shadow_effect(self.format_buttons[fmt], color_hex="#000000", radius=8, x_offset=0, y_offset=3)
        else:
            self.selected_formats.discard(fmt)
            self.format_buttons[fmt].setStyleSheet("QPushButton:hover { background-color: #3498db; }")
            apply_shadow_effect(self.format_buttons[fmt], color_hex="#000000", radius=8, x_offset=0, y_offset=3)

    def add_all_formats(self):
        for fmt, btn in self.format_buttons.items():
            btn.setChecked(True)
            self.toggle_format(fmt, True)

    def remove_all_formats(self):
        for fmt, btn in self.format_buttons.items():
            btn.setChecked(False)
            self.toggle_format(fmt, False)
    # --- END NEW ---

    def update_search_button_state(self):
        """Disables search button if DB is not connected."""
        db_connected = self.db_tab_ref.db is not None
        self.search_button.setEnabled(db_connected)
        if not db_connected:
            self.results_count_label.setText("Not connected to database.")
        else:
            self.results_count_label.setText("Ready to search.")

    def get_selected_tags(self) -> List[str]:
        """Parses comma-separated tags from the QLineEdit."""
        tags_str = self.tags_edit.text().strip()
        if not tags_str:
            return None
        return [tag.strip().lower() for tag in tags_str.split(',') if tag.strip()]

    # --- NEW: Helper to get selected formats ---
    def get_selected_formats(self) -> Optional[List[str]]:
        """Gets the list of selected formats."""
        if self.dropdown:
            if not self.selected_formats:
                return None
            return list(self.selected_formats)
        else:
            formats_str = self.input_formats_edit.text().strip()
            if not formats_str:
                return None
            return [f.strip().lstrip('.').lower() for f in formats_str.replace(',', ' ').split() if f.strip()]
    
    def perform_search(self):
        """Perform the image search using the database."""
        db = self.db_tab_ref.db
        if not db:
            QMessageBox.warning(self, "Error", "Please connect to the database first.")
            return

        self.search_button.setEnabled(False)
        self.search_button.setText("Searching...")
        QApplication.processEvents()
    
        self.clear_results()
        
        # Get search criteria
        group = self.group_combo.currentText().strip() or None
        subgroup = self.subgroup_combo.currentText().strip() or None
        filename = self.filename_edit.text().strip() or None
        tags = self.get_selected_tags()
        formats = self.get_selected_formats() # NEW
        
        if not group and not subgroup and not filename and not tags and not formats: # MODIFIED
            self.results_count_label.setText("Please enter at least one search criterion.")
            self._reset_search_button()
            return
        
        try:
            # Perform database search
            matching_files = db.search_images(
                group_name=group,
                subgroup_name=subgroup,
                tags=tags,
                filename_pattern=filename,
                input_formats=formats, # NEW
                limit=100 
            )
            
            self.display_results(matching_files)

        except Exception as e:
            QMessageBox.critical(self, "Search Error", f"An error occurred during search:\n{str(e)}")
            self.results_count_label.setText(f"Error: {str(e)}")
        
        QTimer.singleShot(200, self._reset_search_button)
    
    def _reset_search_button(self):
        """Reset search button to original style"""
        self.search_button.setEnabled(True)
        self.search_button.setText("Search Database")
    
    def display_results(self, results: List[Dict[str, Any]]):
        """Display the search results (dictionaries) as thumbnails"""
        count = len(results)
        self.results_count_label.setText(f"Found {count} matching image(s)")
        
        if count == 0:
            return
        
        columns = 4 
        for i, img_data in enumerate(results):
            row = i // columns
            col = i % columns
            
            file_path = img_data.get('file_path')
            
            result_container = QWidget()
            result_layout = QVBoxLayout(result_container)
            result_layout.setContentsMargins(5, 5, 5, 5)
            
            image_label = QLabel()
            image_label.setFixedSize(150, 150)
            image_label.setAlignment(Qt.AlignCenter)
            image_label.setStyleSheet("border: 1px solid #4f545c; background: #36393f;")

            pixmap = QPixmap(str(file_path))
            if not pixmap.isNull():
                scaled_pixmap = pixmap.scaled(150, 150, Qt.KeepAspectRatio, Qt.SmoothTransformation)
                image_label.setPixmap(scaled_pixmap)
            else:
                image_label.setText("Not Found")
            
            result_layout.addWidget(image_label)
            
            filename_label = QLabel(img_data.get('filename', 'N/A'))
            filename_label.setWordWrap(True)
            filename_label.setAlignment(Qt.AlignCenter)
            filename_label.setStyleSheet("font-size: 10px;")
            result_layout.addWidget(filename_label)
            
            btn_layout = QHBoxLayout()
            
            view_button = QPushButton("View")
            apply_shadow_effect(view_button, color_hex="#000000", radius=8, x_offset=0, y_offset=3)
            view_button.clicked.connect(lambda chk, p=file_path: self.open_file_preview(p))
            btn_layout.addWidget(view_button)
            
            folder_button = QPushButton("Folder")
            apply_shadow_effect(folder_button, color_hex="#000000", radius=8, x_offset=0, y_offset=3)
            folder_button.clicked.connect(lambda chk, p=file_path: self.open_file_directory(p))
            btn_layout.addWidget(folder_button)
            
            result_layout.addLayout(btn_layout)
            
            self.results_layout.addWidget(result_container, row, col)
            self.result_widgets.append(result_container)

    def remove_preview_window(self, window_instance: ImagePreviewWindow):
        """Removes a preview window from the tracking list when it's closed."""
        try:
            if window_instance in self.open_preview_windows:
                self.open_preview_windows.remove(window_instance)
        except (RuntimeError, ValueError):
            pass

    def open_file_preview(self, file_path: str):
        """Opens the full-size image preview window."""
        if not file_path or not os.path.exists(file_path) or not os.path.isfile(file_path):
            QMessageBox.warning(self, "Invalid Path", f"File not found at path:\n{file_path}")
            return

        for window in self.open_preview_windows:
            if window.image_path == file_path:
                window.activateWindow() 
                return
        
        preview = ImagePreviewWindow(file_path, self.db_tab_ref, parent=self) 
        preview.finished.connect(lambda result, p=preview: self.remove_preview_window(p))
        preview.show() 
        self.open_preview_windows.append(preview)

    def open_file_directory(self, file_path: str):
        """Open the file's parent directory with the default system application"""
        if not file_path or not os.path.exists(file_path):
            QMessageBox.warning(self, "Invalid Path", f"File not found at path:\n{file_path}")
            return
            
        directory = os.path.dirname(file_path)
        system = platform.system()
        try:
            if system == "Windows":
                os.startfile(directory)
            elif system == "Darwin":  # macOS
                subprocess.run(['open', directory])
            else:  # Linux
                subprocess.run(['xdg-open', directory])
        except Exception as e:
            QMessageBox.critical(self, "Error", f"Failed to open directory:\n{e}")
    
    def clear_results(self):
        """Clear all previous search results"""
        for widget in self.result_widgets:
            widget.deleteLater()
        self.result_widgets.clear()
        
        for window in self.open_preview_windows[:]:
            window.close()
        self.open_preview_windows.clear()
    
    def collect(self) -> Dict[str, Any]:
        """Collect search parameters"""
        return {
            "group_name": self.group_combo.currentText().strip() or None,
            "subgroup_name": self.subgroup_combo.currentText().strip() or None, 
            "filename_pattern": self.filename_edit.text().strip() or None,
            "input_formats": self.get_selected_formats() or None, # NEW
            "tags": self.get_selected_tags() or None
        }

File: app/src/gui/tabs/SearchTab.cpp
#include "tabs/SearchTab.h"

// Assumed paths
#include "tabs/DatabaseTab.h"
#include "src/core/PgvectorImageDatabase.h"
#include "gui/components/OptionalField.h"
#include "gui/components/ImagePreviewWindow.h"
#include "gui/styles/Style.h" // for apply_shadow_effect
#include "src/utils/Definitions.h" // for SUPPORTED_IMG_FORMATS

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QGridLayout>
#include <QScrollArea>
#include <QGroupBox>
#include <QFormLayout>
#include <QLineEdit>
#include <QPushButton>
#include <QComboBox>
#include <QLabel>
#include <QMessageBox>
#include <QTimer>
#include <QApplication>
#include <QPixmap>
#include <QDesktopServices>
#include <QUrl>
#include <QFileInfo>

// Assuming SUPPORTED_IMG_FORMATS is a QStringList from definitions.h
// extern const QStringList SUPPORTED_IMG_FORMATS;

SearchTab::SearchTab(DatabaseTab* db_tab_ref, bool dropdown, QWidget *parent)
    : BaseTab(parent),
      m_db_tab_ref(db_tab_ref),
      m_dropdown(dropdown),
      m_filename_field(nullptr),
      m_formats_field(nullptr),
      m_input_formats_edit(nullptr)
{
    init_ui();
    update_search_button_state();
}

SearchTab::~SearchTab()
{
    // Close all open preview windows when the tab is destroyed
    clear_results();
}

void SearchTab::init_ui()
{
    QVBoxLayout *layout = new QVBoxLayout(this);
    
    QGroupBox *search_group = new QGroupBox("Search Database");
    QFormLayout *form_layout = new QFormLayout(search_group);
    form_layout->setContentsMargins(10, 20, 10, 10);
    
    group_combo = new QComboBox;
    group_combo->setEditable(true);
    group_combo->setPlaceholderText("e.g., Summer Trip (Optional)");
    form_layout->addRow("Group name:", group_combo);
    
    subgroup_combo = new QComboBox;
    subgroup_combo->setEditable(true);
    subgroup_combo->setPlaceholderText("e.g., Beach Photos (Optional)");
    form_layout->addRow("Subgroup name:", subgroup_combo);
    
    m_filename_edit = new QLineEdit;
    m_filename_edit->setPlaceholderText("e.g., *.png, img_001, etc (Optional)");
    m_filename_field = new OptionalField("Filename pattern", m_filename_edit, false, this);
    form_layout->addRow(m_filename_field);
    
    if (m_dropdown) {
        QVBoxLayout *formats_layout = new QVBoxLayout;
        QHBoxLayout *btn_layout = new QHBoxLayout;
        for (const QString &fmt : SUPPORTED_IMG_FORMATS) {
            QPushButton *btn = new QPushButton(fmt);
            btn->setCheckable(true);
            btn->setStyleSheet("QPushButton:hover { background-color: #3498db; }");
            style::apply_shadow_effect(btn);
            connect(btn, &QPushButton::clicked, this, [this, fmt, btn](){ this->toggle_format(fmt, btn->isChecked()); });
            btn_layout->addWidget(btn);
            m_format_buttons[fmt] = btn;
        }
        formats_layout->addLayout(btn_layout);

        QHBoxLayout *all_btn_layout = new QHBoxLayout;
        m_btn_add_all = new QPushButton("Add All");
        m_btn_add_all->setStyleSheet("background-color: green; color: white;");
        style::apply_shadow_effect(m_btn_add_all);
        connect(m_btn_add_all, &QPushButton::clicked, this, &SearchTab::add_all_formats);
        m_btn_remove_all = new QPushButton("Remove All");
        m_btn_remove_all->setStyleSheet("background-color: red; color: white;");
        style::apply_shadow_effect(m_btn_remove_all);
        connect(m_btn_remove_all, &QPushButton::clicked, this, &SearchTab::remove_all_formats);
        all_btn_layout->addWidget(m_btn_add_all);
        all_btn_layout->addWidget(m_btn_remove_all);
        formats_layout->addLayout(all_btn_layout);

        QWidget *formats_container = new QWidget;
        formats_container->setLayout(formats_layout);
        m_formats_field = new OptionalField("Input formats", formats_container, false, this);
        form_layout->addRow(m_formats_field);
    } else {
        m_input_formats_edit = new QLineEdit;
        m_input_formats_edit->setPlaceholderText("e.g. jpg png gif (optional)");
        form_layout->addRow("Input formats:", m_input_formats_edit);
    }

    m_tags_edit = new QLineEdit;
    m_tags_edit->setPlaceholderText("tag1, tag2, tag3... (comma-separated, optional)");
    form_layout->addRow("Tags:", m_tags_edit);
    layout->addWidget(search_group);
    
    m_search_button = new QPushButton("Search Database");
    m_search_button->setStyleSheet(R"(
        QPushButton {
            background: qlineargradient(x1:0, y1:0, x2:0, y2:1, stop:0 #667eea, stop:1 #764ba2);
            color: white; font-weight: bold; font-size: 16px;
            padding: 14px; border-radius: 10px; min-height: 44px;
        }
        QPushButton:hover { background: qlineargradient(x1:0, y1:0, x2:0, y2:1, stop:0 #764ba2, stop:1 #667eea); }
        QPushButton:disabled { background: #4f545c; color: #a0a0a0; }
        QPushButton:pressed { background: #5a67d8; }
    )");
    style::apply_shadow_effect(m_search_button);
    connect(m_search_button, &QPushButton::clicked, this, &SearchTab::perform_search);
    layout->addWidget(m_search_button);
    
    connect(group_combo->lineEdit(), &QLineEdit::returnPressed, m_search_button, &QPushButton::click);
    connect(subgroup_combo->lineEdit(), &QLineEdit::returnPressed, m_search_button, &QPushButton::click);
    connect(m_filename_edit, &QLineEdit::returnPressed, m_search_button, &QPushButton::click);
    connect(m_tags_edit, &QLineEdit::returnPressed, m_search_button, &QPushButton::click);
    if (m_input_formats_edit) {
        connect(m_input_formats_edit, &QLineEdit::returnPressed, m_search_button, &QPushButton::click);
    }
    
    layout->addWidget(new QLabel("Search Results:"));
    m_results_count_label = new QLabel("Not connected to database.");
    m_results_count_label->setStyleSheet("color: #aaa; font-style: italic;");
    layout->addWidget(m_results_count_label);
    
    m_results_scroll = new QScrollArea;
    m_results_scroll->setWidgetResizable(true);
    m_results_scroll->setMinimumHeight(300);
    
    m_results_widget = new QWidget;
    m_results_layout = new QGridLayout(m_results_widget);
    m_results_layout->setAlignment(Qt::AlignTop | Qt::AlignHCenter);
    m_results_scroll->setWidget(m_results_widget);
    
    layout->addWidget(m_results_scroll);
}

// --- Format Toggling ---
void SearchTab::toggle_format(const QString& fmt, bool checked)
{
    if (checked) {
        m_selected_formats.insert(fmt);
        m_format_buttons[fmt]->setStyleSheet(R"(
            QPushButton:checked { background-color: #3320b5; color: white; }
            QPushButton:hover { background-color: #00838a; }
        )");
        style::apply_shadow_effect(m_format_buttons[fmt]);
    } else {
        m_selected_formats.remove(fmt);
        m_format_buttons[fmt]->setStyleSheet("QPushButton:hover { background-color: #3498db; }");
        style::apply_shadow_effect(m_format_buttons[fmt]);
    }
}

void SearchTab::add_all_formats()
{
    for (auto it = m_format_buttons.begin(); it != m_format_buttons.end(); ++it) {
        it.value()->setChecked(true);
        toggle_format(it.key(), true);
    }
}

void SearchTab::remove_all_formats()
{
    for (auto it = m_format_buttons.begin(); it != m_format_buttons.end(); ++it) {
        it.value()->setChecked(false);
        toggle_format(it.key(), false);
    }
}

// --- Search Logic ---
void SearchTab::update_search_button_state()
{
    bool db_connected = m_db_tab_ref && m_db_tab_ref->db != nullptr;
    m_search_button->setEnabled(db_connected);
    if (!db_connected) {
        m_results_count_label->setText("Not connected to database.");
    } else {
        m_results_count_label->setText("Ready to search.");
    }
}

QStringList SearchTab::get_selected_tags()
{
    return join_list_str(m_tags_edit->text());
}

QStringList SearchTab::get_selected_formats()
{
    QStringList formats;
    if (m_dropdown) {
        if (m_selected_formats.isEmpty()) return QStringList();
        formats = QStringList(m_selected_formats.values());
    } else {
        formats = join_list_str(m_input_formats_edit->text());
    }
    
    for (QString &f : formats) {
        f = f.remove(0, 1); // remove leading '.'
    }
    return formats;
}

void SearchTab::perform_search()
{
    if (!m_db_tab_ref || !m_db_tab_ref->db) {
        QMessageBox::warning(this, "Error", "Please connect to the database first.");
        return;
    }
    auto* db = m_db_tab_ref->db;

    m_search_button->setEnabled(False);
    m_search_button->setText("Searching...");
    QApplication::processEvents();
    
    clear_results();
    
    QString group = group_combo->currentText().trimmed();
    QString subgroup = subgroup_combo->currentText().trimmed();
    QString filename = m_filename_edit->text().trimmed();
    QStringList tags = get_selected_tags();
    QStringList formats = get_selected_formats();
    
    if (group.isEmpty() && subgroup.isEmpty() && filename.isEmpty() && tags.isEmpty() && formats.isEmpty()) {
        m_results_count_label->setText("Please enter at least one search criterion.");
        _reset_search_button();
        return;
    }
    
    try {
        QList<QJsonObject> matching_files = db->search_images(
            group.isEmpty() ? QString() : group,
            subgroup.isEmpty() ? QString() : subgroup,
            tags.isEmpty() ? QStringList() : tags,
            filename.isEmpty() ? QString() : filename,
            formats.isEmpty() ? QStringList() : formats,
            100
        );
        display_results(matching_files);

    } catch (const std::exception& e) {
        QMessageBox::critical(this, "Search Error", "An error occurred during search:\n" + QString(e.what()));
        m_results_count_label->setText(QString("Error: %1").arg(e.what()));
    }
    
    QTimer::singleShot(200, this, &SearchTab::_reset_search_button);
}

void SearchTab::_reset_search_button()
{
    m_search_button->setEnabled(true);
    m_search_button->setText("Search Database");
}

// --- Result Handling ---
void SearchTab::display_results(const QList<QJsonObject>& results)
{
    int count = results.length();
    m_results_count_label->setText(QString("Found %1 matching image(s)").arg(count));
    if (count == 0) return;
    
    int columns = 4;
    for (int i = 0; i < count; ++i) {
        QJsonObject img_data = results[i];
        int row = i / columns;
        int col = i % columns;
        
        QString file_path = img_data.value("file_path").toString();
        
        QWidget *result_container = new QWidget;
        QVBoxLayout *result_layout = new QVBoxLayout(result_container);
        result_layout->setContentsMargins(5, 5, 5, 5);
        
        QLabel *image_label = new QLabel;
        image_label->setFixedSize(150, 150);
        image_label->setAlignment(Qt::AlignCenter);
        image_label->setStyleSheet("border: 1px solid #4f545c; background: #36393f;");

        QPixmap pixmap(file_path);
        if (!pixmap.isNull()) {
            image_label->setPixmap(pixmap.scaled(150, 150, Qt::KeepAspectRatio, Qt::SmoothTransformation));
        } else {
            image_label->setText("Not Found");
        }
        result_layout->addWidget(image_label);
        
        QLabel *filename_label = new QLabel(img_data.value("filename").toString("N/A"));
        filename_label->setWordWrap(true);
        filename_label->setAlignment(Qt::AlignCenter);
        filename_label->setStyleSheet("font-size: 10px;");
        result_layout->addWidget(filename_label);
        
        QHBoxLayout *btn_layout = new QHBoxLayout;
        QPushButton *view_button = new QPushButton("View");
        style::apply_shadow_effect(view_button);
        connect(view_button, &QPushButton::clicked, this, [this, file_path](){ this->open_file_preview(file_path); });
        btn_layout->addWidget(view_button);
        
        QPushButton *folder_button = new QPushButton("Folder");
        style::apply_shadow_effect(folder_button);
        connect(folder_button, &QPushButton::clicked, this, [this, file_path](){ this->open_file_directory(file_path); });
        btn_layout->addWidget(folder_button);
        
        result_layout->addLayout(btn_layout);
        
        m_results_layout->addWidget(result_container, row, col);
        m_result_widgets.append(result_container);
    }
}

void SearchTab::remove_preview_window(ImagePreviewWindow* window_instance)
{
    m_open_preview_windows.removeOne(window_instance);
}

void SearchTab::open_file_preview(const QString& file_path)
{
    if (file_path.isEmpty() || !QFileInfo::exists(file_path) || !QFileInfo(file_path).isFile()) {
        QMessageBox::warning(this, "Invalid Path", "File not found at path:\n" + file_path);
        return;
    }

    for (ImagePreviewWindow* window : m_open_preview_windows) {
        if (window->image_path == file_path) {
            window->activateWindow();
            return;
        }
    }
    
    ImagePreviewWindow *preview = new ImagePreviewWindow(file_path, m_db_tab_ref, this);
    connect(preview, &ImagePreviewWindow::finished, this, [this, preview](){ this->remove_preview_window(preview); });
    preview->show();
    m_open_preview_windows.append(preview);
}

void SearchTab::open_file_directory(const QString& file_path)
{
    if (file_path.isEmpty() || !QFileInfo::exists(file_path)) {
        QMessageBox::warning(this, "Invalid Path", "File not found at path:\n" + file_path);
        return;
    }
    QString directory = QFileInfo(file_path).absolutePath();
    QDesktopServices::openUrl(QUrl::fromLocalFile(directory));
}

void SearchTab::clear_results()
{
    for (QWidget *widget : m_result_widgets) {
        widget->deleteLater();
    }
    m_result_widgets.clear();
    
    for (ImagePreviewWindow *window : m_open_preview_windows) {
        window->close();
    }
    m_open_preview_windows.clear();
}

// --- BaseTab Overrides ---
void SearchTab::browse_files() {}
void SearchTab::browse_directory() {}
void SearchTab::browse_input() {}
void SearchTab::browse_output() {}

QJsonObject SearchTab::collect()
{
    QJsonObject out;
    out["group_name"] = group_combo->currentText().trimmed();
    out["subgroup_name"] = subgroup_combo->currentText().trimmed();
    out["filename_pattern"] = m_filename_edit->text().trimmed();
    
    QStringList formats = get_selected_formats();
    if (!formats.isEmpty()) {
        out["input_formats"] = QJsonArray::fromStringList(formats);
    }
    
    QStringList tags = get_selected_tags();
    if (!tags.isEmpty()) {
        out["tags"] = QJsonArray::fromStringList(tags);
    }
    return out;
}